### PR TITLE
adding Bard handler 

### LINKED ISSUE #8500: add support for Google Bard API
Add support for Google Bard API

### PR DESCRIPTION
**Pull request by someshfengde**
## Description

added bard handler to do querying with google gemini-pro model 
Fixes #8500 

## Type of change
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [x] 📄 This change requires a documentation update

## Additional Media:

- [x] I have attached a brief loom video or screenshots showcasing the new functionality or change.
image

## Checklist:
- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.

### CODE REVIEW

**Comment by someshfengde** [comment]
@paxcema  @dusvyat  can you please check this PR ?

File: mindsdb/integrations/handlers/bard_handler/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Bard Handler
+The Bard ML handler integrates the Bard library from Google's Generative AI with MindsDB. You can use it to generate text completions with the Gemini Pro model for your existing text data.
+
+## Google Generative AI (Bard)
+Google Generative AI is a library that provides access to powerful language models for text generation. The Bard library, used in this handler, offers various models for generating text. More information about the Google Generative AI library can be found [here](https://github.com/GoogleCloudPlatform/generative-ai).
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/README.md:5]: I don't see a Bard library in the link. Furthermore, if the underlying integration is to Gemini models I would request that you please rename this integration to "google_gemini_handler" (don't worry about the original issue, we can modify it afterwards).

File: mindsdb/integrations/handlers/bard_handler/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+*Note:* Ensure you have the necessary API key for accessing the Bard library.
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/README.md:7]: Please explain the difference between the API key required at ML engine creation time versus the one passed when creating the actual model. In the screenshot attached to the PR description you use two different API keys (which, by the way, you did not blur and so these are now compromised).

It is not clear what each one does. If possible, perhaps rename them to be more descriptive (e.g. `cloud_api_key` and `model_api_key`)

File: mindsdb/integrations/handlers/bard_handler/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+```sql
+SELECT t.question, m.answer
+FROM mindsdb.bard_test as m
+JOIN files.question_table as t;
+```
+
+# Bard Handler Python Code Explanation
+
+The provided Python code defines a `BardHandler` class, which is a subclass of `BaseMLEngine` in MindsDB's integration library. Here's a breakdown of its key components:
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/README.md:44]: This reads like it was AI generated, and it is not useful when compared to the base ML engine tutorial in our docs. Please remove the entire section or rephrase it to describe any important details, as opposed to the general structure.

File: mindsdb/integrations/handlers/bard_handler/__about__.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+__title__ = "MindsDB Bard handler"
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/__about__.py:1]: We're now asking contributors to ignore adding `__about__` and `__setup__` files as they will not be needed moving forward, so please remove this one.

File: mindsdb/integrations/handlers/bard_handler/__init__.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+from mindsdb.integrations.libs.const import HANDLER_TYPE
+from mindsdb.utilities import log
+
+logger = log.getLogger(__name__)
+
+from .__about__ import __description__ as description
+from .__about__ import __version__ as version
+
+try:
+    from .bard_handler import BardHandler as Handler
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/__init__.py:10]: As discussed above, please rename the handler to avoid using `Bard` because it is not accurate.

File: mindsdb/integrations/handlers/bard_handler/bard_handler.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+import google.generativeai as genai
+import pandas as pd
+from mindsdb.integrations.libs.base import BaseMLEngine
+from mindsdb.utilities import log
+from mindsdb.utilities.config import Config
+
+logger = log.getLogger(__name__)
+
+
+class BardHandler(BaseMLEngine):
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/bard_handler.py:13]: As discussed above, please rename the handler to avoid using Bard because it is not accurate.

File: mindsdb/integrations/handlers/bard_handler/bard_handler.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Integration with the Anthropic LLM Python Library
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/bard_handler.py:15]: Reference to Anthropic instead of Google, please fix.

File: mindsdb/integrations/handlers/bard_handler/bard_handler.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    """
+
+    name = "bard"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.default_chat_model = "gemini-pro"
+        self.supported_chat_models = ["gemini-pro"]
+        # self.default_max_tokens = 100
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/bard_handler.py:24]: If this is not needed, then please remove it.

File: mindsdb/integrations/handlers/bard_handler/bard_handler.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+    def predict_answer(self, text):
+        """
+        connects with bard api to predict the answer for the particular question
+
+        """
+
+        args = self.model_storage.json_get("args")
+        if "ai_prompt" not in args["using"]:
+            args["using"]["ai_prompt"] = ""
```
Comment (paxcema) [inline mindsdb/integrations/handlers/bard_handler/bard_handler.py:94]: Could you explain the reasoning behind the following prompt template?

```f"{HUMAN_PROMPT} {text} {AI_PROMPT}"```

I'm not sure I follow.

Assuming it is what we want, additionallyy are these empty prompts good default values? Seems like it could lead to poor model completions.

**Review by paxcema — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/bard_handler/README.md (hunk 1 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/README.md:5]: Yes sure will make the change

File: mindsdb/integrations/handlers/bard_handler/README.md (hunk 3 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/README.md:44]: yes will do

File: mindsdb/integrations/handlers/bard_handler/__about__.py (hunk 4 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/__about__.py:1]: ohh I see 👍

File: mindsdb/integrations/handlers/bard_handler/__init__.py (hunk 5 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/__init__.py:10]: yes sure

File: mindsdb/integrations/handlers/bard_handler/bard_handler.py (hunk 6 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/bard_handler.py:13]: yes sure

File: mindsdb/integrations/handlers/bard_handler/bard_handler.py (hunk 8 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/bard_handler.py:24]: Yes sure 
but in future it might be needed ( there are some more models by google genai API but they are not yet available )

File: mindsdb/integrations/handlers/bard_handler/bard_handler.py (hunk 9 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/bard_handler.py:94]: I referred the anthropic handler while creating this one that's why I included this

File: mindsdb/integrations/handlers/bard_handler/README.md (hunk 2 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/README.md:7]: yes will revoke that key

Comment (someshfengde) [reply mindsdb/integrations/handlers/bard_handler/README.md:7]: btw both are the same keys don't have any difference

**Comment by someshfengde** [comment]
hey @paxcema I resolved the mentioned comments

**Review by paxcema — CHANGES_REQUESTED** [review]
Looking great! One last thing we'd need before approving this is testing. Please add a test for the handler so that we can run it and programmatically check it works both now and in the future when we introduce any changes.

**Comment by someshfengde** [comment]
Yes sure will do it asap

**Comment by someshfengde** [comment]
added some tests

File: tests/unit/ml_handlers/test_google_gemini.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+                column='question',
+                model='non-existing-gemini-model',
+                api_key='{GEMINI_API_KEY}';
+            """
+        )
+        with pytest.raises(Exception):
+            self.wait_predictor("proj", "test_google_invalid_model")
+
+    @pytest.mark.skip(reason="This test is failing as no error is being thrown")
+    def test_unknown_model_argument(self):
```
Comment (dusvyat) [inline tests/unit/ml_handlers/test_google_gemini.py:37]: @someshfengde This test fails. the reason being there is no argument checker in the handler. I marked to skip until we fix

I suggest you use `pydantic`  to validate handler arguments on create - see [ here](https://github.com/mindsdb/mindsdb/blob/61f389ac7e5843459544b9860b5085229e800bd4/mindsdb/integrations/handlers/rag_handler/rag_handler.py#L61l) for an example

File: mindsdb/integrations/handlers/google_gemini_handler/google_gemini_handler.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        self,
+        target: str,
+        df: Optional[pd.DataFrame] = None,
+        args: Optional[Dict] = None,
+    ) -> None:
+        if "model" not in args["using"]:
+            args["using"]["model"] = self.default_chat_model
+        elif args["using"]["model"] not in self.supported_chat_models:
+            raise Exception(
+                f"Invalid chat model. Please use one of {self.supported_chat_models}"
```
Comment (dusvyat) [inline mindsdb/integrations/handlers/google_gemini_handler/google_gemini_handler.py:37]: May be worth checking api key is valid in create

**Review by dusvyat — CHANGES_REQUESTED** [review]
@someshfengde Many thankd for your contribution.

A few minor comments to address before merging.

File: mindsdb/integrations/handlers/google_gemini_handler/google_gemini_handler.py (hunk 11 above)
Comment (someshfengde) [reply mindsdb/integrations/handlers/google_gemini_handler/google_gemini_handler.py:37]: yes sure

File: tests/unit/ml_handlers/test_google_gemini.py (hunk 10 above)
Comment (someshfengde) [reply tests/unit/ml_handlers/test_google_gemini.py:37]: I had added this into the create method previously i think this should be alright 

https://github.com/mindsdb/mindsdb/blob/d22b743e733d1475202ea4970e726fd68a1e66bb/mindsdb/integrations/handlers/google_gemini_handler/google_gemini_handler.py#L35C1-L35C1

**Comment by someshfengde** [comment]
Added the api-key validation into the create method also the supported models is already there. 

Not sure if I have to add the new validation from pydantic 

https://github.com/mindsdb/mindsdb/blob/d22b743e733d1475202ea4970e726fd68a1e66bb/mindsdb/integrations/handlers/google_gemini_handler/google_gemini_handler.py#L35C1-L35C1

**Review by dusvyat — APPROVED** [review]

**Comment by dusvyat** [comment]
@someshfengde looks good, we can skip `pydantic` part for now. This gets us most of the way there. Thanks again for your contribution.
